addons/binding/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/handler/KeContactP20Handler.java

```
@@ -0,0 +1,967 @@
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonParser;
+
+/**
+ * The {@link KeContactP20Handler} is responsible for handling commands, which
+ * are sent to one of the channels.
+ * 
+ * @author Karel Goderis - Initial contribution
+ */
+public class KeContactP20Handler extends BaseThingHandler {
+
+	public static final String IP_ADDRESS = "ipAddress";
+	public static final String POLLING_REFRESH_INTERVAL = "refreshInterval";
+
+	public static final int CONNECTION_REFRESH_INTERVAL = 100;
+	public static final int REPORT_INTERVAL = 2000;
+	public static final int PING_TIME_OUT = 3000;
+	public static final int BUFFER_SIZE = 1024;
+	public static final int REMOTE_PORT_NUMBER = 7090;
+	public static final int LISTENER_PORT_NUMBER = 7090;
+
+	private Logger logger = LoggerFactory.getLogger(KeContactP20Handler.class);
+
+	private Selector selector;
+	private DatagramChannel datagramChannel = null;
+	protected SelectionKey datagramChannelKey = null;
+	protected DatagramChannel listenerChannel = null;
+	protected SelectionKey listenerKey = null;
+	private final Lock lock = new ReentrantLock();
+
+	private ScheduledFuture<?> connectionJob;
+	private ScheduledFuture<?> pollingJob;
+
+	private int maxPresetCurrent = 0;
+	private int maxSystemCurrent = 63000;
+
+	public KeContactP20Handler(Thing thing) {
+		super(thing);
+	}
+
+	@Override
+	public void initialize() {
+		logger.debug("Initializing KEBA KeContact P20 handler.");
+
+		try {
+			selector = Selector.open();
+		} catch (IOException e) {
+			logger.error(
```

Is this an unrecoverable error situation? Then you should actually throw a runtime exception here.